vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+
+pub mod nv12;
+
+pub const ACTORS_COUNT: usize = 11;
+
+pub type Migrator<BS> = Arc<dyn ActorMigration<BS> + Send + Sync>;
+pub type MigrationResult<T> = Result<T, MigrationError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    // FIXME: use underlying concrete types when possible.
+    #[error("Failed creating job for state migration: {0}")]
+    MigrationJobCreate(String),
+    #[error("Failed running job for state migration: {0}")]
+    MigrationJobRun(String),
+    #[error("Flush failed post migration: {0}")]
+    FlushFailed(String),
+    #[error("Failed writing to blockstore: {0}")]
+    BlockStoreWrite(String),
+    #[error("Failed reading from blockstore: {0}")]
+    BlockStoreRead(String),
+    #[error("Migrator not found for cid: {0}")]
+    MigratorNotFound(Cid),
+    #[error("Failed updating new actor state: {0}")]
+    SetActorState(String),
+    #[error("State tree creation failed")]
+    StateTreeCreation(String),
+    #[error("Incomplete migration specification with {0} code CIDs")]
+    IncompleteMigrationSpec(usize),
+    #[error("Thread pool creation failed: {0}")]
+    ThreadPoolCreation(ThreadPoolBuildError),
+    #[error("Migration failed")]
+    Other,
+}
+
+pub struct StateMigration<BS> {
+    migrations: HashMap<Cid, Migrator<BS>>,
+    deferred_code_ids: HashSet<Cid>,
+}
+
+impl<BS: BlockStore + Send + Sync> StateMigration<BS> {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            migrations: HashMap::new(),
+            deferred_code_ids: HashSet::new(),
+        }
+    }
+
+    pub fn add_migrator(&mut self, prior_cid: Cid, migrator: Migrator<BS>) {
+        self.migrations.insert(prior_cid, migrator);
+    }
+
+    pub fn migrate_state_tree(
+        &mut self,
+        store: Arc<BS>,
+        prior_epoch: ChainEpoch,
+        actors_in: StateTree<BS>,
+        mut actors_out: StateTree<BS>,
+    ) -> MigrationResult<Cid> {
+        if self.migrations.len() + self.deferred_code_ids.len() != ACTORS_COUNT {
+            return Err(MigrationError::IncompleteMigrationSpec(
+                self.migrations.len(),
+            ));
+        }
+
+        let cpus = num_cpus::get();
+        let chan_size = cpus / 2;
+
+        log::info!(
+            "Using {} CPUs for migration and channel size of {}",
+            cpus,
+            chan_size
+        );
+
+        let pool = rayon::ThreadPoolBuilder::new()
+            .thread_name(|id| format!("nv12 migration thread: {}", id))
+            .num_threads(cpus)
+            .build()
+            .map_err(MigrationError::ThreadPoolCreation)?;
+
+        let (state_tx, state_rx) = crossbeam_channel::bounded(chan_size);
+        let (job_tx, job_rx) = crossbeam_channel::bounded(chan_size);
+
+        pool.scope(|s| {
+            let store_clone = store.clone();
+
+            s.spawn(move |_| {
+                actors_in
+                    .for_each(|addr, state| {
```

Comment:
It would be nice if `StateTree` just implemented `Iterator` instead of this `for_each` method which returns a `Result<(), E>`. That way you could implement your own error handling, or lack thereof since this lambda isn't fallible.

Reply:
I think it would require quite a bit of refactoring to get this to implement Iterator sadly

Reply:
Just speculating, but I think it would just require adding a `.iter()` or `.into_iter()` before the `for_each` and changing the `for_each` to `try_for_each`.

Reply:
Everywhere that currently uses the custom `for_each` and wants to continue iterating with a fallible closure.

Reply:
StateTree doesn't implement `Iterator`. If it implements that, the Hamt type would also need to implement that, where the Node type contained within would also need to implement that. So it's probably a non-trivial change and better tackled as a separate PR. I'll add a TODO for that.